philander/stc311_reg.py

```diff
@@ -0,0 +1,360 @@
+
+from enum import auto
+from .systypes import Enum
+
+
+class ChipType(Enum):
+    STC3115 = auto()
+    STC3117 = auto()
+
+
+class _ModeValues:
+    """This type describes different operating modes.
+    """
+    MODE_VMODE = 0x01  # 0: Mixed mode (Coulomb counter active); 1: Power saving voltage mode
+    MODE_ALM_ENA = 0x08  # Alarm function enable
+    MODE_GG_RUN = 0x10  # Standby / operating mode
+    MODE_FORCE_CC = 0x20  # Forces the relaxation timer to switch to the Coulomb counter (CC) state.
+    MODE_FORCE_VM = 0x40  # Forces the relaxation timer to switch to voltage mode (VM) state.
+
+
+class _CtrlValues:
+    """This type describes different control commands.
+    """
+    CTRL_IO0DATA = 0x01  # ALM pin status / ALM pin output drive
+    CTRL_GG_RST = 0x02  # resets the conversion counter GG_RST is a self-clearing bit
+    CTRL_GG_VM = 0x04  # Coulomb counter mode / Voltage mode active
+    CTRL_BATFAIL = 0x08  # Battery removal (BATD high).
+    CTRL_PORDET = 0x10  # Power on reset (POR) detection / Soft reset
+    CTRL_ALM_SOC = 0x20  # Set with a low-SOC condition
+    CTRL_ALM_VOLT = 0x40  # Set with a low-voltage condition
+    CTRL_DEFAULT = CTRL_IO0DATA
+
+
+class _STC311x_Reg:
+    def __init__(self, Param_dict):
+        # apply config
+        self.CONFIG_GASGAUGE_0_RSENSE = Param_dict["Gasgauge.senseResistor"]
+        self.CONFIG_GASGAUGE_0_BATTERY_IDX = Param_dict["Gasgauge.battery_idx"]  # TODO: what is this pin used for?
+        self.CONFIG_GASGAUGE_0_GPIO_ALARM = Param_dict["Gasgauge.gpio_alarm_idx"]
+        self.RELAX_MAX_DEFAULT = Param_dict["Gasgauge.cmonit_max"]
+        self.SETUP_0_REG_CC_CNF = Param_dict["Gasgauge.cc_cnf"]
+        self.SETUP_0_REG_VM_CNF = Param_dict["Gasgauge.vm_cnf"]
+        self.SETUP_0_REG_ALARM_SOC = Param_dict["Gasgauge.alarm_soc"]
+        self.SETUP_0_REG_ALARM_VOLTAGE = Param_dict["Gasgauge.alarm_voltage"]
+        self.SETUP_0_REG_CURRENT_THRES = Param_dict["Gasgauge.current_thres"]
+        self.SETUP_0_REG_CMONIT_MAX = Param_dict["Gasgauge.cmonit_max"]
+
+    # Definition of registers and their content
+
+    REG_MODE = 0  # Mode register
+    # see MODE_VALUES for possible values
+
+    REG_CTRL = 1  # Control and status register
+    # see CTRL_VALUES for possible values
+
+    REG_SOC_L = 2
+    REG_SOC_H = 3
+    REG_SOC = REG_SOC_L  # Battery state of charge (SOC)
+    REG_COUNTER_L = 4
+    REG_COUNTER_H = 5
+    REG_COUNTER = REG_COUNTER_L  # Number of conversions
+    REG_CURRENT_L = 6
+    REG_CURRENT_H = 7
+    REG_CURRENT = REG_CURRENT_L  # Battery current, voltage drop over sense resistor
+    REG_VOLTAGE_L = 8
+    REG_VOLTAGE_H = 9
+    REG_VOLTAGE = REG_VOLTAGE_L  # Battery voltage
+    REG_TEMPERATURE = 10  # Temperature [C]
+    # REG 11, 12 see __init__()
+    REG_OCV_L = 13
+    REG_OCV_H = 14
+    REG_OCV = REG_OCV_L  # OCV register
+    REG_CC_CNF_L = 15
+    REG_CC_CNF_H = 16
+    REG_CC_CNF = REG_CC_CNF_L  # Coulomb counter gas gauge configuration
```

Comment:
I have not found a definitive solution for setting defaults yet, and still respecting the given parameters. Either the parameter must be None by default and in the case the default from the REG class is used, or the default value is not stored in the REG class at all, but only in the default parameters set defined in Params_init (I personally like the idea to have a singular set of defaults more, that way the user can call params_init on an empty dict and receive all defaults. Otherwise they would only receive a lot of None values, not knowing the actual default without looking in the code or docs respectively.)

Reply:
It's just about adding a line like
`CC_CNF_DEFAULT = 395`
So we have a generic handle to what the default/reset value of this register is. It may or may not be used, lateron.